src/XAD/Hessian.hpp

```diff
@@ -50,9 +51,10 @@ class Hessian
         : foo_(func), v_(v), domain_(static_cast<unsigned int>(v_.size())), matrix_(0)
     {
         if (std::distance(first, last) != domain_)
-        {
-            // throw exception
-        }
+            throw OutOfRange("Iterator not allocated enough space");
+        static_assert(has_begin<decltype(*first)>::value,
```

Comment:
```typename std::iterator_traits<RowIterator>::value_type```

Reply:
This is just an additional check in the template to make sure `RowIterator` resolves to an iterator type correct? Or is something wrong with the static_assert/exception I need to fix.

Reply:
This is a more portable and generic way to check the value type of the iterator. It is to replace `decltype(*first)`.

Reply:
Ah thank you, I see how that's better.